### PR TITLE
chore: Ensure time column is projected in subqueries

### PR DESCRIPTION
**Pull request by stuartcarnie**
Helps #6891

The goal of this PR is to ensure the `SelectStatement` AST is rewritten to include the `time` column in the root `SELECT` and any subqueries. It also ensures the `time` column is projected to the first index when formatted as InfluxQL output.

Another key change of this PR is to remove the dependency on IOx `schema::Schema` and its associated metadata, which will not be available when processing subqueries.

### CODE REVIEW

File: iox_query_influxql/src/plan/util.rs [hunk 1]
```diff
@@ -25,29 +24,25 @@ pub(in crate::plan) fn binary_operator_to_df_operator(op: BinaryOperator) -> Ope
...
-            iox_schema: schema_from_df(df_schema)?,
         })
     }
+
+    /// Returns `true` if the field `name` is a tag type.
+    pub(super) fn is_tag_field(&self, name: &str) -> bool {
+        self.df_schema
+            .fields()
+            .iter()
+            .any(|f| f.name() == name && matches!(f.data_type(), DataType::Dictionary(..)))
```
Comment (crepererum) [inline iox_query_influxql/src/plan/util.rs:44]: This is a bit risky if we ever wanna:

- extend our field data type support (why not encode fields as dictionary if they're low-cardinality?)
- change the encoding for keys (e.g. use ordinary string cols for high-cardinality tags or use RLE)

So I would prefer if we could either use the IOx schema directly or if we could use the schema metadata to detect this (in the end this is what the IOx wrapper does as well).

**Review by wolffcm — COMMENTED** [review]
The changes to projecting the `time` column seem good to me. 

I have similar concerns as Marco though, I wonder if this is locking us into the current representations of tag as dictionary and string field as utf8.

> Another key change of this PR is to remove the dependency on IOx schema::Schema and its associated metadata, which will not be available when processing subqueries.

It's not clear to me why this metadata is not available in subqueries. Can you explain?

**Comment by stuartcarnie** [comment]
@wolffcm / @crepererum 

Regarding IOx schema metadata.

I wanted it to work that way too, but the metadata is lost through the logical plans, and is some cases for good reason. The reason I encoded this heuristic into a function `is_tag_field` was so that we could evolve it if such a situation arises.

## Background

I've explored this problem of using IOx schema extensively and it had been discussed it at length in Slack and other tickets that the schema metadata is not propagated through logical plan (and physical plan) nodes. @alamb has suggested it would be considerable work for this to be plumbed through DataFusion, but it turns out there are still cases the metadata cannot be propagated as it would be non-sensical.

The direct issue I ran into is that you can construct an IOx schema from a `TableScan` node, but any other nodes, and the metadata is lost so this function always fails:

https://github.com/influxdata/influxdb_iox/blob/abe5d26d2f52d660c9a1ea17923ea5e9172fa5be/schema/src/lib.rs#L167

As it is unable to extract the field metadata:

https://github.com/influxdata/influxdb_iox/blob/abe5d26d2f52d660c9a1ea17923ea5e9172fa5be/schema/src/lib.rs#L181-L187

### Why is the field metadata lost?

Many of the functions to build logical plans simply drop the field metadata. For example, say we wish to fetch the current schema for a node, which happens to be an `Aggregate` node. We fetch the `DFSchema` via the `schema` call of the `LogicalPlan` node:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/logical_plan/plan.rs#L133

This returns the `schema` associated with this node. `Aggregate` nodes are constructed using the `aggregate` builder function:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/logical_plan/builder.rs#L845

which calls `Aggregate::try_new`:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/logical_plan/builder.rs#L852-L856

`try_new` constructs the schema for the Aggregate node:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/logical_plan/plan.rs#L1559-L1562

and clones the schema-level metadata, but the fields (or columns) of the schema are constructed via the call to `exprlist_to_fields`:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/utils.rs#L986

and as this is an Aggregate node, will resolve the `fields` via a call to `exprlist_to_fields_aggregate`:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/utils.rs#L996-L998

The `exprlist_to_fields_aggregate` ultimately calls the `to_field` method:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/utils.rs#L977

which constructs the new `DFField` without any source metadata:

https://github.com/apache/arrow-datafusion/blob/5d3802fba0578448937e654475d667f9e410daad/datafusion/expr/src/expr_schema.rs#L275-L285

> **Note**
>
> This is quite valid, as we can imagine that there are many cases these field are not projecting the underlying column, but combined as some more complex expression or aggregate, so there is no valid IOx metadata.

The `to_field` function is called when constructing a `Projection` node, `WindowAggregate` and many others.

### Example

An example of why we can expect the metadata to be unavailable, would be a query like the following:

```sql
SELECT usage_idle * 5 as foo FROM cpu
```

As we might expect, this logical plan is a `TableScan` followed by a `Projection` of the expression `usage_idle * 5` as column alias `foo`. When retrieving the DFSchema for the `Projection` plan node, it makes sense the `foo` column no longer has IOx metadata, as it is not strictly a field.

It becomes complicated for subqueries, which is the reason I had to move away from using it directly:

```sql
SELECT foo, bar FROM (SELECT some_tag as foo, usage_idle * 5 as bar FROM cpu)
```

The column `foo` no longer carries the tag metadata when calling the schema of the subquery plan node. What would happen if we allowed tags to be combined as an expression in the future?

I would welcome a brainstorm for alternatives if anyone is available?

File: influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected [hunk 2]
```diff
@@ -70,17 +70,43 @@ name: m0
...
+| time                | f64  | tag0  |
++---------------------+------+-------+
+| 2022-10-31T02:00:00 | 10.1 | val00 |
+| 2022-10-31T02:00:00 | 11.3 | val01 |
+| 2022-10-31T02:00:00 | 10.4 | val02 |
+| 2022-10-31T02:00:10 | 21.2 | val00 |
+| 2022-10-31T02:00:10 | 18.9 | val00 |
+| 2022-10-31T02:00:20 | 11.2 | val00 |
+| 2022-10-31T02:00:30 | 19.2 | val00 |
++---------------------+------+-------+
```
Comment (NGA-TRAN) [inline influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:83]: Question: Do we want time column to be always shown first no matter what order it is selected?
The other columns in the order of their selection?

**Review by wolffcm — APPROVED** [review]
Given the additional context (thanks @stuartcarnie) I think this is quite reasonable.

File: iox_query_influxql/src/plan/util.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/util.rs:44]: Agree, so I have created #7739 as a follow on to this to address that by introducing an intermediate state when processing InfluxQL `SELECT` statements.

**Comment by stuartcarnie** [comment]
@crepererum / @wolffcm 

I think the only clean way of resolving this is to create a new, intermediate state for `SELECT` statements, which I have pondered a few times. That state is produced after rewriting and normalising a `SELECT` statement and expanding things such as wild cards. Furthermore, this intermediate representation can contain state specific to planning, such as resolved schema, fields, etc. It will simplify a lot of the planner code, as there will be a lot of "internal" errors that go away for state that is no longer possible. I'll plan to do that as a follow up per #7739

**Pull request merged**